src/main/java/com/epam/aidial/core/storage/BlobWriteStream.java

```diff
@@ -99,19 +106,16 @@ public void end(Handler<AsyncResult<Void>> handler) {
                     throw new RuntimeException(exception);
                 }
 
-                Buffer lastChunk = chunkBuffer.slice(0, position);
-                metadata = new FileMetadata(resource, bytesHandled, contentType);
+                byte[] lastChunk = chunkBuffer.slice(0, position).getBytes();
                 if (mpu == null) {
-                    log.info("Resource is too small for multipart upload, sending as a regular blob");
-                    storage.store(resource.getAbsoluteFilePath(), contentType, lastChunk);
+                    bytes = lastChunk;
                 } else {
                     if (position != 0) {
-                        MultipartPart part = storage.storeMultipartPart(mpu, ++chunkNumber, lastChunk);
+                        MultipartPart part = blobStorage.storeMultipartPart(mpu, ++chunkNumber, lastChunk);
                         parts.add(part);
                     }
-
-                    storage.completeMultipartUpload(mpu, parts);
-                    log.info("Multipart upload committed, bytes handled {}", bytesHandled);
+                    String newEtag = etagBuilder.append(lastChunk).build();
+                    mpu.blobMetadata().getUserMetadata().put(ResourceUtil.ETAG_ATTRIBUTE, newEtag);
```

Comment:
Where do you complete?

Reply:
Partially reverted, now completes in same place.